### DISCUSSION TITLE
Question about the Implementation of quantize Method which calls prune and dedup

### DISCUSSION THREAD

**Discussion by gz65555** [donmccurdy/glTF-Transform#1179] [Q&A] [answered]
I've noticed that the `quantize` method implementation involves calls to both `prune` and `dedup` functions. My understanding was that `quantize` is primarily intended to reduce the file size by limiting the precision of vertex attributes. However, the inclusion of `prune` and `dedup` seems to suggest additional functionality.

I have a model where I intend to use the `quantize` method to compress the data, but I do not wish to remove duplicate materials, which is the current behavior due to the `dedup` function being called within quantize.

**Comment by donmccurdy** [comment]
Quantization requires modifying data throughout the scene, sometimes including materials. To clean up any unneeded data, `quantize` partially runs the `prune` and `dedup` functions.

Probably the cleanup could be made more precise - to deduplicate materials only if materials were modified, for example - but that isn't implemented currently.

See:

https://github.com/donmccurdy/glTF-Transform/blob/f261f4b4536526f373096762b88a9c87d954813e/packages/functions/src/quantize.ts#L127-L134

**Comment by gz65555** [comment]
@donmccurdy thank you for explaining. Perhaps, consider adding more specific options to control the behavior of `prune` and `dedup`. For instance, 

```typescript
doc.transform(quantize({
 dedup: false,
 prune: false
}));
```

**Reply by donmccurdy** [reply]
Could I ask why you want duplicates of materials in the model? 

Disabling these steps means quantize would add redundant data to your file – possibly quite a lot. I think we would need a workaround that removes any extra data added by the quantize step _without_ removing extra data that might have been there before. Please feel free to open a feature request or PR for this – I'm open to the idea, but I don't have plans to work on it myself right now.

Another workaround would be to make a locally modified copy of the quantize() function and run that in your pipeline instead.

**Reply by gz65555** [reply]
Well, the real-world scenario is that designers have a series of models, and we optimize the material size through quantization. These materials of glTF have the same properties in glTF, so they are removed. However, we actually need to adjust these materials at runtime to achieve different effects in different locations.

**Reply by gz65555** [reply]
This is the issue: https://github.com/donmccurdy/glTF-Transform/issues/1185. We can discuss solutions under this issue, and I'll open a pull request over the weekend if I have time.

**Reply by donmccurdy** [reply, marked as answer]
Resolved in https://github.com/donmccurdy/glTF-Transform/issues/1185.
